### DISCUSSION TITLE
Can you use multiple problem_matchers or multiple copies of "on_output_parse"?

### DISCUSSION THREAD

**Discussion by braxtons12** [stevearc/overseer.nvim#108] [Q&A] [answered]
The build system I use for c++ prints `error: ` before it starts the compiler's error messages.

This means that a single-pass problem_matcher won't work correctly for this build system (If I match on "error: file:line:col:sev:text" I only get the first error, if I match on "file:line:col:sev:text" I only get errors 2 thru N).

The natural/easy solution to this would just be to either use two problem_matchers or two copies of the "on_output_parse" component, but from what I can tell so far it looks like neither of those options is supported. Is this correct?

**Comment by stevearc** [comment, marked as answer]
Check the docs on parsers https://github.com/stevearc/overseer.nvim/blob/master/doc/parsers.md
You can specify patterns using regular expressions, so you should be able to do something like `^(error: )?` at the start to match "error:" if it's present and pass if it isn't. If you can't figure out how to get a problem matcher or a parser that works for your output, paste a few lines here as a sample and I can help

**Reply by braxtons12** [reply]
So here's an example output from my build system:

```
error: src/serde/serialize/Json.cpp:12:5: error: expected unqualified-id
    }
    ^
src/serde/serialize/Json.cpp:17:5: error: expected unqualified-id
    }
    ^
src/serde/serialize/Json.cpp:22:5: error: expected unqualified-id
    }
```

and I know this lua pattern will match the error from the first line

```
lua_pat = [[^error: ([^:]+):(%d+):(%d+):%s+([^:]+):%s+(.*)$]],
```

and this one will match the second and third errors

```
lua_pat = [[^([^:]+):(%d+):(%d+):%s+([^:]+):%s+(.*)$]],
```

But I need a pattern or regex that will work for both

**Reply by stevearc** [reply]
This is where lua patterns show their limitations and regular expressions really shine. Here is the a problemMatcher for you:

```lua
{
  pattern = {
    vim_regexp = "\\v^(error: )?([^:]+):(\\d+):(\\d+): ([^:]+): (.*)$",
    kind = "location",
    file = 2,
    line = 3,
    column = 4,
    severity = 5,
    message = 6,
}}
```
Note the optional capture group at the beginning `(error: )?`. It will match that text if present, but it's optional. Since it _does_ capture the text, we start the numbering of the fields at 2 since the filename will technically be the second capture group in this expression.

To use it, pass this to the `on_output_parse` component as demonstrated in [the docs on problem matchers](https://github.com/stevearc/overseer.nvim/blob/master/doc/parsers.md#problem-matchers).

I don't expect you to have figured this out on your own, but the way that I built this was by following the process outlined in [writing parsers](https://github.com/stevearc/overseer.nvim/blob/master/doc/parsers.md#writing-parsers). I pasted your output in the bottom panel, and in the parser panel I put

```lua
local parser = require("overseer.parser")
local problem_matcher = require("overseer.template.vscode.problem_matcher")

local pm = problem_matcher.resolve_problem_matcher({
  pattern = {
    vim_regexp = "\\v^(error: )?([^:]+):(\\d+):(\\d+): ([^:]+): (.*)$",
    kind = "location",
    file = 2,
    line = 3,
    column = 4,
    severity = 5,
    message = 6,
}})
local parser_defn = problem_matcher.get_parser_from_problem_matcher(pm)

return parser.new({ diagnostics = parser_defn })
```

That framework allowed me to tweak the regular expression and make sure it was working.

**Reply by braxtons12** [reply]
Awesome, this works. Thanks so much!
